Bulk bugfixes 2

NBKelly opened the pull request:
Going through a few more old issues.
* Lewi/lovegood was fixed at some point, so I added a test.
* I added an implementation for Zed 1.0/2.0 that crawls through the run log, and finds out if any break abilities that cost :lose-click have been used. The implementation looks ugly, I would appreciate advice on how to make it nicer.
* I wanted to make aumakua better, so I tried using `:end-breach-server`, only to find that it tracked literally nothing except the server that was breached.  So I added some state tracking for each breach, and that information is returned with `:end-breach-server`. Right now, aumakua should work with divide and conquer, gang sign, kitsune, shiro, etc, properly now. I'm going to add some tests for this soon. Once that's done, this closes #6355 and closes #4455
* I expanded breach tracking enough that we can fix mwanza city grid (closes #5616, closes #6253, closes #3714)

NoahTheDuke commented on src/clj/game/cards/programs.clj:
```diff
@@ -424,10 +424,9 @@
-                                                      (:did-steal target)))
-                                             (:did-access target)))
+                    :events [{:event :end-breach-server
+                              :req (req (and (not (or (:did-steal target)
```
Unnecessary `and` here.

NoahTheDuke commented on src/clj/game/cards/ice.clj:
```diff
@@ -304,6 +304,29 @@
+        cards (vec (map #(first (second %)) breaks))
+        subs (map #(:subroutines %) cards)
+        broken-subs (map #(filter :broken %) subs)
+        breakers (flatten (map #(map :breaker %) broken-subs))
```
```suggestion
        breakers (mapcat #(map :breaker %) broken-subs)
```

NoahTheDuke commented on src/clj/game/cards/ice.clj:
```diff
@@ -304,6 +304,29 @@
+        ;; If Adjusted Matrix ever gets correctly implemented, there will be
+        ;; a minor edge case here if the runner uses a non-click break ab on a
+        ;; card hosting adjusted matrix -nbkelly, 2022
+        actual-breakers (map #(find-cid % (get-all-cards state)) breakers)
```
maybe something like `(let [all-cards (get-all-cards state)] (map #(find-cid % all-cards) breakers))` so we're only calling `get-all-cards` once instead of once per breaker.

NoahTheDuke commented on src/clj/game/cards/ice.clj:
```diff
@@ -304,6 +304,29 @@
+        ;; a minor edge case here if the runner uses a non-click break ab on a
+        ;; card hosting adjusted matrix -nbkelly, 2022
+        actual-breakers (map #(find-cid % (get-all-cards state)) breakers)
+        abs (map #(if (= (:side %) "Runner") (:abilities %) (:runner-abilities %)) actual-breakers)
+        costs (map :break-cost (apply concat abs))
```
```suggestion
        abs (mapcat #(if (= (:side %) "Runner") (:abilities %) (:runner-abilities %)) actual-breakers)
        costs (map :break-cost abs)
```

NoahTheDuke commented on src/clj/game/cards/ice.clj:
```diff
@@ -3561,14 +3584,21 @@
-   :subroutines [(do-brain-damage 1)
-                 (do-brain-damage 1)]
+  {:subroutines [{:label "Do 1 brain damage"
+                  :effect (req (if (spent-click-to-break-sub state run)
+                                 (resolve-ability state side (do-brain-damage 1) card nil)
+                                 (system-msg state side "does not do brain damage with Zed 1.0")))}
```
```suggestion
                  :async true
                  :effect (req (if (spent-click-to-break-sub state run)
                                 (continue-ability state side (do-brain-damage 1) card nil)
                                 (do (system-msg state side "does not do brain damage with Zed 1.0")
                                     (effect-completed state side eid))}
```
Something like this, you'll probably have to clean it up. `do-brain-damage` is an `:async` ability, so it has to be resolved with `continue-ability` and the calling ability has to be `:async` too.

NoahTheDuke replied:
Oh yeah, and this applies to the other Zed 1 and 2 subroutines.

NoahTheDuke commented:
Lookin good, nbkelly! thanks for doing all this work.

NBKelly commented:
I think I tackled all of the suggestions. I also added `from-server` back into `end-breach` (I accidentally took it out before).

NBKelly commented:
The one above should fix all the instances where a single mwanza forces +3/+6/+9... accesses from triggering multiple times. 
I think for the economic ability, we can probably tackle all #5616, #6253 and #3714 by having the 'gain-credits' ability be registered when breaching, instead of just being there forever.
This would also require each breach tracking the cards accessed (which is next on my list of things to do).

NBKelly commented:
The errata text for turtle now states "whenever a breach ends and...", so it's actually incorrect to add counters for accesses which do not breach a server (Such as otoroshi, or Quest Complete).

I think turtle is 100% correctly implemented now, and any issues leftover would happen to be cards which shouldn't breach a server but do, or cards that do but shouldn't.

I've added tests for turtle vs. gang sign, divide & conquer. I'll be done with this PR once I've added tests for mwanza.

NBKelly commented:
I think that caps out all the changes for this one :)
